### PR TITLE
Model CI Release

### PR DESCRIPTION
**Pull request by sasha-scale**
🚢 🚢 🚢  🚀 🚀🚀 Add model CI extension to Nucleus python client 🚀 🚀🚀 🚢 🚢 🚢

### CODE REVIEW

File: nucleus/modelci/data_transfer_objects/eval_function.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+from ...pydantic_base import ImmutableModel
+from ..constants import ThresholdComparison
+
+
+class EvaluationCriterion(ImmutableModel):
+    """
+    eval_function_id: ID of evaluation function
+    threshold_comparison: comparator for evaluation. i.e. threshold=0.5 and threshold_comparator > implies that a test only passes if score > 0.5.
+    threshold: numerical threshold that together with threshold comparison, defines success criteria for test evaluation.
```
Comment (phil-scale) [inline nucleus/modelci/data_transfer_objects/eval_function.py:13]: Does this show up as we want in the autodocs?

File: nucleus/model.py [hunk 2]
```diff
@@ -159,3 +162,28 @@ def create_run(
...
         return model_run
+
+    def evaluate(self, unit_test_names: List[str]) -> AsyncJob:
+        """Evaluates this on the specified Unit Tests. ::
+
+            import nucleus
+            client = nucleus.NucleusClient("YOUR_SCALE_API_KEY")
+            model = client.list_models()[0]
+            unit_test = client.modelci.create_unit_test(
+                "sample_unit_test", "slc_bx86ea222a6g057x4380"
```
Comment (phil-scale) [inline nucleus/model.py:173]:
```suggestion
                "sample_unit_test", "YOUR_SLICE_ID"
```

**Review by phil-scale — COMMENTED** [review]
We should update the Changelog as well.

File: nucleus/__init__.py [hunk 3]
```diff
@@ -882,29 +888,9 @@ def make_request(
...
-            )
-            if response.status_code not in RetryStrategy.statuses:
-                break
-            time.sleep(retry_wait_time)
-
-        if not response.ok:
-            self.handle_bad_response(endpoint, requests_command, response)
-
-        return response.json()
+        if payload is None:
```
Comment (ardila) [inline nucleus/__init__.py:891]: why?

File: nucleus/modelci/__init__.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
+"""Model CI Python Library."""
+from typing import List
+
+from nucleus.connection import Connection
+from nucleus.job import AsyncJob
+
+from .constants import (
```
Comment (ardila) [inline nucleus/modelci/__init__.py:7]: Lets only include the imports in this file, and move the client to its own folder.

So you would just add from .client import ModelCI

File: nucleus/modelci/data_transfer_objects/eval_function.py (hunk 1 above)
Comment (ardila) [reply nucleus/modelci/data_transfer_objects/eval_function.py:13]: Example usage would be helpful in the docstring.

File: nucleus/modelci/eval_functions/available_eval_functions.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+class BoundingBoxPrecision(BaseEvalFunction):
+    @classmethod
+    def expected_name(cls) -> str:
+        return "bbox_precision"
+
+
+class CustomEvalFunction(BaseEvalFunction):
+    @classmethod
+    def expected_name(cls) -> str:
+        return "private_custom_function"  # Placeholder: See super().eval_func_entry for actual name
```
Comment (ardila) [inline nucleus/modelci/eval_functions/available_eval_functions.py:39]: Can this raise an error when not implemented (NotImplementedError?) bit confused about this

File: nucleus/modelci/data_transfer_objects/eval_function.py (hunk 1 above)
Comment (drakejwong) [reply nucleus/modelci/data_transfer_objects/eval_function.py:13]: It does not! https://scale-ai-nucleus-python-sdk--176.com.readthedocs.build/en/176/api/nucleus/modelci/index.html#nucleus.modelci.EvaluationCriterion

We'd need to indent these under a `Parameters:` header and as a best practice, add a oneliner summary of the class. Also because pydantic isn't yet supported by our autodoc framework, we'd need to add types manually.
```suggestion
    <oneline summary>
    
    Parameters:
        eval_function_id (str): ID of evaluation function
        threshold_comparison (:class:`ThresholdComparison`): comparator for evaluation. i.e. threshold=0.5 and threshold_comparator > implies that a test only passes if score > 0.5.
        threshold (float): numerical threshold that together with threshold comparison, defines success criteria for test evaluation.
```

File: nucleus/modelci/unit_test_metric.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
+from nucleus.pydantic_base import ImmutableModel
+
+from .constants import ThresholdComparison
+
+
+class UnitTestMetric(ImmutableModel):
```
Comment (ardila) [inline nucleus/modelci/unit_test_metric.py:6]: When should we use ImmutableModel and when should we use Dataclass?

File: nucleus/modelci/unit_test.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+            import nucleus
+            client = nucleus.NucleusClient("YOUR_SCALE_API_KEY")
+            unit_test = client.modelci.create_unit_test(
+                "sample_unit_test", "slc_bx86ea222a6g057x4380"
+            )
+
+            e = client.modelci.eval_functions
+            # Assuming a user would like to add all available public evaluation functions as criteria
+            unit_test.add_criterion(
+                e.bbox_iou() > 0.5
```
Comment (ardila) [inline nucleus/modelci/unit_test.py:61]: Sorry, can someone explain how this works? I thought that this python syntax would just pass a boolean value to add_criterion. Can the way this works be documented somewhere?

Comment (sasha-scale) [reply nucleus/modelci/unit_test.py:61]: Added a follow up ticket for this: https://app.shortcut.com/scaleai/story/314738/document-use-of-pydantic-immutablemodel-and-evaluationcriterion-syntax-for-posterity - I don't feel like I can speak to this very well and we're planning to ship this tonight

FYI @gatli

**Comment by sasha-scale** [comment]
Draft release notes: 

Adds an extension to the Nucleus Client for newly introduced Model CI functionality.  With the Model CI, extension, users can use Nucleus slice’s to define test cases that describe critical edge case scenarios. A unit test can have one or more evaluation criteria, which define how to apply evaluation functions to a unit test.  The newly introduced classes are:

- UnitTest
- EvaluationFunction
- UnitTest Evaluation
- UnitTestMetric

Documentation can be found at: [Nucleus API Reference](https://docs.nucleus.scale.com/en/latest/)

Happy Coding!

File: CHANGELOG.md [hunk 8]
```diff
@@ -4,6 +4,15 @@ All notable changes to the [Nucleus Python Client](https://github.com/scaleapi/n
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [0.4.0](https://github.com/scaleapi/nucleus-python-client/releases/tag/v0.3.0) - 2021-08-12
```
Comment (phil-scale) [inline CHANGELOG.md:7]: should the link be v0.4.0?

File: nucleus/modelci/__init__.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
+"""Model CI Python Library."""
+
+__all__ = [
+    "ModelCI",
+    "UnitTest",
+]
```
Comment (phil-scale) [inline nucleus/modelci/__init__.py:6]: Curious why `UnitTestEvaluation` and `UnitTestItemEvaluation` aren't here.

Comment (sasha-scale) [reply nucleus/modelci/__init__.py:6]: Screen Shot 2021-12-08 at 10 39 18 PM
My understanding is that it creates these shortcuts at the top of the page! 

We could add more classes later on but by default they're sorted alphabetically which is kind of weird (ModelCI should definitely be at the top).  It's difficult to figure out ordering so I'm inclined to leave that to a follow up PR

Comment (pfmark) [reply nucleus/modelci/__init__.py:6]: We had more initially, but I think they just create confusion for now. We can also add them in, I just thought those two are the most important for interaction. Wdyt @phil-scale ?

**Review by phil-scale — APPROVED** [review]
Sounds good, LGTM!

File: CHANGELOG.md (hunk 8 above)
Comment (gatli) [reply CHANGELOG.md:7]: Yes!

File: nucleus/modelci/unit_test_metric.py (hunk 6 above)
Comment (gatli) [reply nucleus/modelci/unit_test_metric.py:6]: I'd suggest everything over the wire (`data_transfer_objects`) should be a `ImmutableModel` since we benefit from `pydantic` validation. Dataclasses are then used as the domain objects that can be created from the `data_transfer_objects`.

File: nucleus/modelci/unit_test.py (hunk 7 above)
Comment (gatli) [reply nucleus/modelci/unit_test.py:61]: I'll add some docs around this. But the gist of it is that we've overriden the operator methods to define the `EvaluationCriteria`...  like @phil-scale said. 🙂

File: CHANGELOG.md (hunk 8 above)
Comment (gatli) [reply CHANGELOG.md]: Updated it.

**Review by pfmark — APPROVED** [review]
